examples/ecs/computed_states.rs

```diff
@@ -0,0 +1,682 @@
+//!
+//! In this case, we'll be implementing the following pattern:
+//! - The game will start in a `Menu` state, which we can return to with `Esc`
+//! - From there, we can enter the game - where our bevy symbol moves around and changes color
+//! - While in game, we can pause and unpause the game using `Space`
+//! - We can also toggle "Turbo Mode" with the `T` key - where the movement and color changes are all faster. This
+//!   is retained between pauses, but not if we exit to the main menu.
+//!
+//! In addition, we want to enable a "tutorial" mode, which will involve it's own state that is toggled in the main menu.
+//! This will display instructions about movement and turbo mode when in game and unpaused, and instructions on how to unpause when paused.
+//!
+//! To implement this, we will create 2 root-level states: [`AppState`] and [`TutorialState`].
+//! We will then create some computed states that derive from [`AppState`]: [`InGame`] and [`TurboMode`] are marker states implemented
+//! as Zero-Sized Structs (ZSTs), while [`IsPaused`] is an enum with 2 distinct states.
+//! And lastly, we'll add [`Tutorial`], a computed state deriving from [`TutorialState`], [`InGame`] and [`IsPaused`], with 2 distinct
+//! states to display the 2 tutorial texts.
+
+use bevy::prelude::*;
+
+// To begin, we want to define our state objects.
+#[derive(Debug, Clone, Copy, Default, Eq, PartialEq, Hash, States)]
+enum AppState {
+    #[default]
+    Menu,
+    // Unlike in the `states` example, we're adding more data in this
+    // version of our AppState. In this case, we actually have
+    // 4 distinct "InGame" states - unpaused and no turbo, paused and no
+    // turbo, unpaused and turbo and paused and turbo.
+    InGame {
+        paused: bool,
+        turbo: bool,
+    },
+}
+
+// The tutorial state object, on the other hand, is a fairly simple enum.
+#[derive(Debug, Clone, Copy, Default, Eq, PartialEq, Hash, States)]
+enum TutorialState {
+    #[default]
+    Active,
+    Inactive,
+}
+
+// Because we have 4 distinct values of `AppState` that mean we're "InGame", we're going to define
+// a separate "InGame" type and implement `ComputedStates` for it.
+// This allows us to only need to check against one type
+// when otherwise we'd need to check against multiple.
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+struct InGame;
+
+impl ComputedStates for InGame {
+    // Our computed state depends on `AppState`, so we need to specify it as the SourceStates type.
+    type SourceStates = AppState;
+
+    // The compute function takes in the `SourceStates` wrapped in `Option`'s, in this case the complex type
+    // signature actually boils down to `Option<AppState>`, but I wanted to leave the complex version visible
+    // so you are aware of what it is in case it is generated for you by your IDE.
+    fn compute(
+        sources: <<Self as ComputedStates>::SourceStates as bevy_internal::ecs::schedule::StateSet>::OptionalStateSet,
```

Comment:
Examples should be something easily copy paste-able by end users - this extended type annotation is more confusing than helpful IMO.

Reply:
Strong +1. This seems to go against other examples. I think there is a strong likelihood this scares away new users who just want to use computed state.

Reply:
Makes sense! will do.